File: leetcode/divide-and-conquor/go/169.go
package conquor

func simple(nums []int) int {
	cnt := make(map[int]int)
	maxCount, maxNum := 0, 0
	for _, v := range nums {
		cnt[v]++
		if cnt[v] > maxCount {
			maxCount = cnt[v]
			maxNum = v
		}
	}
	return maxNum
}

func majorityElement(nums []int) int {
	count, result := 0, 0
	for _, v := range nums {
		if count == 0 {
			count = 1
			result = v
		} else if v == result {
			count++
		} else {
			count--
		}
	}
	return result
}

File: leetcode/divide-and-conquor/cpp/169.cc
#include <vector>
#include <iterator>
using std::vector;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int count = 0;
        int result = 0;
        for(int v : nums) {
            if (count == 0) {
                count = 1;
                result = v;
            } else {
                v == result ? count++ : count--;
            }
        }
        return result;
    }
};
